server/buildbuddy_server/buildbuddy_server.go

```diff
@@ -87,6 +89,60 @@ func (s *BuildBuddyServer) SearchInvocation(ctx context.Context, req *inpb.Searc
 	return searcher.QueryInvocations(ctx, req)
 }
 
+func (s *BuildBuddyServer) authorizeInvocationUpdate(ctx context.Context, reqCtx *ctxpb.RequestContext, in *tables.Invocation) error {
+	auth := s.env.GetAuthenticator()
+	if auth == nil {
+		return status.UnimplementedError("Not Implemented")
+	}
+	if _, err := auth.AuthenticatedUser(ctx); err != nil {
+		return err
+	}
+
+	// Invocations should probably never be writeable by OTHERS,
+	// but this check is included for completeness.
+	if in.Perms&perms.OTHERS_WRITE != 0 {
+		return nil
+	}
+
+	if reqCtx.GetUserId() == nil {
+		return status.InternalError("User ID should not be nil after authentication.")
+	}
+	isOwner := reqCtx.GetUserId().GetId() == in.UserID
+	if isOwner && in.Perms&perms.OWNER_WRITE != 0 {
+		return nil
+	}
+	isGroupSelected := reqCtx.GetGroupId() == in.GroupID
+	if isGroupSelected && in.Perms&perms.GROUP_WRITE != 0 {
+		return nil
+	}
+
+	return status.PermissionDeniedError("You do not have permission to edit this invocation.")
+}
+
+func (s *BuildBuddyServer) UpdateInvocation(ctx context.Context, req *inpb.UpdateInvocationRequest) (*inpb.UpdateInvocationResponse, error) {
+	db := s.env.GetInvocationDB()
+	if req.GetInvocationId() == "" {
+		return nil, status.InvalidArgumentError("Invocation ID is required.")
+	}
+	updatedPerms, err := perms.ToPerms(req.GetAcl())
+	if err != nil {
+		return nil, status.InvalidArgumentError("The provided ACL is invalid.")
+	}
+	in, err := db.LookupInvocation(ctx, req.GetInvocationId())
+	if err != nil {
+		return nil, err
+	}
+	if err := s.authorizeInvocationUpdate(ctx, req.GetRequestContext(), in); err != nil {
+		return nil, err
+	}
+	// NOTE: We currently ignore the user_id and group_id in the updated ACL.
+	in.Perms = updatedPerms
+	if err := db.InsertOrUpdateInvocation(ctx, in); err != nil {
```

Comment:
One slight problem with the way this is currently written is that if an invocation is in-flight (picture a multi-hour long test run), and a user updates permissions on it, it's possible you'll 1) read data at t=0, 2) a invocation update will happen at t=1, and 3) you'll write the old data from t=0 back to the DB with the new ACLs at t=2.

One simple fix would be to only modify the ACL column when you update ACLs? And even better is doing this in a transaction so you're sure the modification happens atomically. I know this is probably a pain -- lmk if you want to chat about the best way to fix it.